spring-cloud-gcp-logging/src/main/java/org/springframework/cloud/gcp/logging/TraceIdLoggingWebMvcInterceptor.java

```diff
@@ -0,0 +1,71 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.logging;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.cloud.logging.TraceLoggingEnhancer;
+
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+/**
+ * {@link org.springframework.web.servlet.HandlerInterceptor} that extracts the request
+ * trace ID from the "x-cloud-trace-context" HTTP header and stores it in a thread-local
+ * using {@link TraceLoggingEnhancer#setCurrentTraceId}.
+ *
+ * The {@link TraceLoggingEnhancer} can then be used in a logging appender to add the
+ * trace ID metadata to log messages.
+ *
+ * @author Mike Eltsufin
+ */
+public class TraceIdLoggingWebMvcInterceptor extends HandlerInterceptorAdapter {
+
+	private static final String X_CLOUD_TRACE = "x-cloud-trace-context";
+
+	@Override
+	public boolean preHandle(HttpServletRequest req,
+			HttpServletResponse resp, Object handler) throws Exception {
+		TraceLoggingEnhancer.setCurrentTraceId(extractTraceIdFromRequest(req));
```

Comment:
Why would one set to the `ThreadLocal` null value?

Take a look into the `ThreadLocal` logic, please:
```
public void set(T value) {
        Thread t = Thread.currentThread();
        ThreadLocalMap map = getMap(t);
        if (map != null)
            map.set(this, value);
        else
            createMap(t, value);
    }
```

So, if there is no a `Map` associated with the current thread, it is created independently of the `value`. Therefore when it is `null` we create extra object for nothing.

Reply:
As I understand, the map entry will be created anyway on the first call to `ThreadLocal.get()`. So, we might as well pay the price upfront.
I could put a guard for `null` if you still think it's worth it.

Reply:
Right, and that is really painful for Java per se.

But from here, in this code, you can't guarantee that someone always is going to call `ThreadLocal.get()` downstream 😄

Reply:
In any case I don't think it's appropriate to address the issue of `null` trace IDs here because TraceLoggingEnhancer already expects the possibility of null trace ID:
```  @Override
  public void enhanceLogEntry(com.google.cloud.logging.LogEntry.Builder builder) {
    String traceId = getCurrentTraceId();
    if (traceId != null) {
      builder.addLabel(traceIdLabel, traceId);
    }
  }```
Though we might have a different opinion on the efficiency impact of that it doesn't feel right to circumvent the logging team's implementation.

Reply:
Created issue: https://github.com/GoogleCloudPlatform/google-cloud-java/issues/2746